tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc

```diff
@@ -224,6 +224,296 @@ static bool HasCanonicalDimensionNumbers(
+  // loop
+  for (int batch_ind = 0; batch_ind < batch_rank; batch_ind++)
+    loop_dim.push_back(std::to_string(batch_ind));
+  // Add the dimension after ellipsis into the loop
+  int post_elip = pre_elip + kEllipsis.size();
+  for (int post_elip_ind = post_elip; post_elip_ind < loop.size();
+       post_elip_ind++)
+    loop_dim.push_back(std::string(1, loop[post_elip_ind]));
+  return loop_dim;
+}
+
+// Check that all operand's batch has same rank
+bool CheckBatchHasEqualRank(const size_t& lhs_rank, StringRef lhs_loop,
+                            const size_t& rhs_rank, StringRef rhs_loop,
+                            const size_t& out_rank, StringRef out_loop) {
+  const std::string kEllipsis = "...";
+  SmallVector<int, 3> batch_rank_vec;
+  if (lhs_rank != lhs_loop.size()) {
+    size_t lhs_batch_rank = lhs_rank - (lhs_loop.size() - kEllipsis.size());
+    batch_rank_vec.push_back(lhs_batch_rank);
+  }
+  if (rhs_rank != rhs_loop.size()) {
+    size_t rhs_batch_rank = rhs_rank - (rhs_loop.size() - kEllipsis.size());
+    batch_rank_vec.push_back(rhs_batch_rank);
+  }
+  if (out_rank != out_loop.size()) {
+    size_t out_batch_rank = out_rank - (out_loop.size() - kEllipsis.size());
+    batch_rank_vec.push_back(out_batch_rank);
+  }
+  if (!std::equal(batch_rank_vec.begin() + 1, batch_rank_vec.end(),
+                  batch_rank_vec.begin()) &&
+      batch_rank_vec.size() > 1)
+    return false;
+  return true;
+}
+// Convert mhlo.einsum op into linalg.generic
+// Algorithm in general 3 steps:
+
+// Step1) Dissect entire einsum_config to different operands
+// e.g f("abc,cd->abd") = {lhs:["abc"], rhs:["cd"], out:["abd"]}
+
+// Step2) Split up the string into vector of the elements
+// e.g {lhs:["abc"], rhs:["cd"], out:["abd"]} = {lhs:["a","b","c"],
+// rhs:["c","d"], out:["a","b","d"]}
+
+// Step3) Convert the vector into data access
+// patern represented by affineMaps with affineDimensions e.g
+// {lhs:["a","b","c"], rhs:["c","d"], out:["a","b","d"]} = {lhs:[d0,d1,d2],
+// rhs:[d2,d3], out:[d0,d1,d3]}
+
+// Note: Step2 is especially important to for batch
+// dim representation using ellipsis, S.T it has no limits to number of
+// dimension that the ellipsis can represent
+class EinsumToLinalgConverter : public OpConversionPattern<mhlo::EinsumOp> {
+ public:
+  using OpConversionPattern<mhlo::EinsumOp>::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      mhlo::EinsumOp op, ArrayRef<Value> args,
+      ConversionPatternRewriter& rewriter) const final {
+    mhlo::EinsumOp::Adaptor adaptor(args);
+    auto get_rank = [](Value v) {
+      return v.getType().cast<ShapedType>().getRank();
+    };
+    auto is_scalar = [&](Value v) { return get_rank(v) == 0; };
+    auto einsum_config = op.einsum_config();
+
+    // Setting Up Commonly Used Symbols as strings
+    const std::string kArrow = "->";
+    const std::string kComma = ",";
+    const std::string kEllipsis = "...";
```

Comment:
I think we can move above utils to private functions and share the constants. We define kEllipsis in many places.

Reply:
done, but this requires me to set some of the functions as static member of the class. Should I put in the rest of the util functions as member of the class too?